src/app/server/CommissioningWindowManager.cpp

```diff
@@ -0,0 +1,276 @@
+        ChipLogError(Discovery, "Operational advertising enabled");
+    }
+}
+
+void CommissioningWindowManager::Cleanup()
+{
+    mServer->GetExchangManager().UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::PBKDFParamRequest);
+    StopAdvertisement();
+
+    // reset all advertising
+    app::MdnsServer::Instance().StartServer(Mdns::CommissioningMode::kDisabled);
+}
+
+void CommissioningWindowManager::OnSessionEstablishmentError(CHIP_ERROR err)
+{
+    Cleanup();
+
+    ChipLogProgress(AppServer, "OnSessionEstablishmentError: %s", ErrorStr(err));
+    ChipLogProgress(AppServer, "Failed in SPAKE2+ handshake");
+
+    if (mAppDelegate != nullptr)
+    {
+        mAppDelegate->OnRendezvousStopped();
+    }
+}
+
+void CommissioningWindowManager::OnSessionEstablished()
+{
+    CHIP_ERROR err = mServer->GetSecureSessionManager().NewPairing(
+        Optional<Transport::PeerAddress>::Value(mPairingSession.GetPeerAddress()), mPairingSession.GetPeerNodeId(),
+        &mPairingSession, SecureSession::SessionRole::kResponder, 0);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Ble, "Failed in setting up secure channel: err %s", ErrorStr(err));
+        OnSessionEstablishmentError(err);
+        return;
+    }
+
+    ChipLogProgress(AppServer, "Device completed SPAKE2+ handshake");
+    if (mAppDelegate != nullptr)
+    {
+        mAppDelegate->OnRendezvousStarted();
+    }
+
+    DeviceLayer::PlatformMgr().AddEventHandler(OnPlatformEventWrapper, reinterpret_cast<intptr_t>(this));
+
+    Cleanup();
+    ChipLogProgress(AppServer, "Device completed Rendezvous process");
+}
+
+CHIP_ERROR CommissioningWindowManager::PrepareCommissioningWindow(uint16_t commissioningTimeoutSeconds,
+                                                                  uint16_t & allocatedSessionID)
+{
+    ReturnErrorOnFailure(mIDAllocator->Allocate(allocatedSessionID));
+
+    ReturnErrorOnFailure(mPairingSession.MessageDispatch().Init(&mServer->GetSecureSessionManager()));
+
+    if (commissioningTimeoutSeconds != kNoCommissioningTimeout)
+    {
+        ReturnErrorOnFailure(
+            DeviceLayer::SystemLayer().StartTimer(commissioningTimeoutSeconds * 1000, HandleCommissioningWindowTimeout, this));
+    }
+
+    ReturnErrorOnFailure(mServer->GetExchangManager().RegisterUnsolicitedMessageHandlerForType(
+        Protocols::SecureChannel::MsgType::PBKDFParamRequest, &mPairingSession));
+
+    ReturnErrorOnFailure(StartAdvertisement());
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CommissioningWindowManager::OpenBasicCommissioningWindow(ResetFabrics resetFabrics, uint16_t commissioningTimeoutSeconds,
+                                                                    CommissioningWindowAdvertisement advertisementMode)
+{
+    RestoreDiscriminator();
+
+#if CONFIG_NETWORK_LAYER_BLE
+    SetBLE(advertisementMode == chip::CommissioningWindowAdvertisement::kBle);
+#else
+    SetBLE(false);
```

Comment:
if no ble layer, should we return an invalid argument if the advertisement mode is BLE?

we could add that check in the `#if` only and let the SetBLE outside the if (since we will know that if no ble layer, thte advertising mode will not be ble.

Reply:
Tried doing this, but looks like the current behavior is this.
1. `OpenBasicCommissioningWindow()` API's `advertisementMode` argument defaults to `chip::CommissioningWindowAdvertisement::kBle`.
2. Most of the device implementations are not providing a value for this argument. So, it uses the default.
3. If the device doesn't support BLE, it doesn't error out, and uses non-BLE commissioning.

Likely all our test devices do support BLE layer. However, some tests and utilities are disabling it. I just tried running `all-cluster-app` locally (with adding this check) and it failed to start.

I think we need to clean up this API, but maybe in a separate PR, as it'll add changes in about 20 additional files. I'll add a TODO here to keep track of this change.

Reply:
Can we change the default based on the availability of ble layer?

A default of 'use ble' when no ble network layer exists seems silly.